esp-hal/src/gpio/interconnect.rs

```diff
@@ -0,0 +1,384 @@
+        }
+    }
+}
+
+/// Multiple pins can be connected to one output signal.
+pub struct OutputSignal {
+    pin: AnyPin,
+    is_inverted: bool,
+}
+
+impl Peripheral for OutputSignal {
+    type P = Self;
+
+    unsafe fn clone_unchecked(&mut self) -> Self::P {
+        Self {
+            pin: self.pin.clone_unchecked(),
+            is_inverted: self.is_inverted,
+        }
+    }
+}
+
+impl Sealed for OutputSignal {}
+
+impl OutputSignal {
+    pub(crate) fn new(pin: AnyPin) -> Self {
+        Self {
+            is_inverted: false,
+            pin,
+        }
+    }
+
+    /// Inverts the peripheral's output signal.
+    pub fn invert(&mut self) {
+        self.is_inverted = !self.is_inverted;
+    }
+
+    /// Inverts the peripheral's input signal.
+    pub fn inverted(mut self) -> Self {
+        self.invert();
+        self
+    }
+
+    fn connect(
+        &self,
+        signal: OutputSignalType,
+        invert: bool,
+        invert_enable: bool,
+        enable_from_gpio: bool,
+        output: u8,
+    ) {
+        unsafe { &*GPIO::PTR }
+            .func_out_sel_cfg(output as usize)
+            .modify(|_, w| unsafe {
+                w.out_sel()
+                    .bits(signal)
+                    .inv_sel()
+                    .bit(invert)
+                    .oen_sel()
+                    .bit(enable_from_gpio)
+                    .oen_inv_sel()
+                    .bit(invert_enable)
+            });
+    }
+}
+
+impl PeripheralOutputPin for OutputSignal {
+    /// Connect the pin to a peripheral output signal.
+    // TODO: the following options should be part of the struct:
+    /// invert: Configures whether or not to invert the output value
+    ///
+    /// invert_enable: Configures whether or not to invert the output enable
+    /// signal
+    ///
+    /// enable_from_gpio: Configures to select the source of output enable
+    /// signal.
+    /// - false = Use output enable signal from peripheral
+    /// - true = Force the output enable signal to be sourced from bit n of
+    ///   GPIO_ENABLE_REG
+    ///
+    /// force_via_gpio_mux: if true don't use the alternate function even if it
+    /// matches
+    fn connect_peripheral_to_output(&mut self, signal: gpio::OutputSignal, _: private::Internal) {
+        let af = if self.is_inverted {
+            GPIO_FUNCTION
+        } else if let Some(af) = self
+            .pin
+            .output_signals(private::Internal)
+            .into_iter()
+            .position(|s| s == Some(signal))
+        {
+            match af {
+                0 => AlternateFunction::Function0,
+                1 => AlternateFunction::Function1,
+                2 => AlternateFunction::Function2,
+                3 => AlternateFunction::Function3,
+                4 => AlternateFunction::Function4,
+                5 => AlternateFunction::Function5,
+                _ => unreachable!(),
+            }
+        } else {
+            GPIO_FUNCTION
+        };
```

Comment:
Idk if this is any more readable.
```Rust
let functions = [
    AlternateFunction::Function0,
    AlternateFunction::Function1,
    AlternateFunction::Function2,
    AlternateFunction::Function3,
    AlternateFunction::Function4,
    AlternateFunction::Function5,
];

self.pin
    .output_signals(private::Internal)
    .into_iter()
    .zip(functions.iter())
    .find(|(s, _)| s == Some(signal))
    .map(|(_, f)| f)
    .unwrap_or(GPIO_FUNCTION);
```

Reply:
Will add `TryFrom<usize>` to `AlternateFunction`